### PR TITLE
remove folding and unfolding of batch dim and sequence dim in model.py

### PR DESCRIPTION
**Pull request by tianyu-l**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #190

Local experiments show that this does not slow down training. Actually it brings slight memory saving and speedup.
Experiment setup: devgpu, dp-degree=2, tp-degree=4, llama-13b, full AC

Before:
Screenshot 2024-04-03 at 3 46 15 PM

After this PR:
Screenshot 2024-04-03 at 4 01 01 PM

### CODE REVIEW

**Comment by awgu** [comment]
Discussing with @tianyu-l offline still: this PR means that we have extra cat after all-gather and reduce-scatter since we are no longer gathering/reducing on dim-0.

**Comment by fduwjj** [comment]
@awgu I am also curious why adding extra cat here would make perf better. I remembered in the past, when doing fold and unfolding, we did see perf gain since all-gather and reduce-scatter is performing on dim-0 which makes use to use all-gather-base and reduce-scatter-base. Have we checked the profiler to see what collectives we are calling @tianyu-l ?

**Comment by tianyu-l** [comment]
After discussing with @awgu, I think the perf testing here is not convincing. Looking at the profile traces right now I do see it introduces extra cat.

Although, combining #314 with the no-folding way, we turn the beginning embedding communication from an all-reduce to a reduce-scatter. But as the number of layers grow, this save should matter less and less.

The PR was mainly to make distributed techniques not intruding model code, with the understanding that we probably should do it even with minor perf regression.

cc: @fduwjj

**Comment by awgu** [comment]
@tianyu-l Is it possible for the folding and unfolding to be part of the `SequenceParallel` API as an option (somehow)?

**Comment by tianyu-l** [comment]
@awgu SequenceParallel is at layer level, I think we need to do the folding and unfolding outside all the layers, to avoid doing it num_of_layers times.

**Comment by fduwjj** [comment]
@tianyu-l I think what you said:

> But as the number of layers grow, this save should matter less and less

This folding-unfolding happens every layer (for every atten and mlp layer) basically for every all-gather that happens for non-0 dim and then if torch.cat is called. we will see this perf save. 

To anwser @awgu's ask, maybe this can be done at c10d level. I remember we used to let all_gather_into_tensor to handle it. But it has correctness issue. Because when a 3-D tensor is sharded by dim 1 for example, then techinically if you all-gather by dim 1 and cat by dim 1 the result is different compared to if we do folding -> all-gather by dim 0 -> unfolding. It's actually very surprising that this works for LLaMa training.

**Comment by tianyu-l** [comment]
update: created #437 to showcase how to do folding and unfolding to accelerate Sequence Parallel, with no plan to land it (added a pointer in #438)

**Comment by tianyu-l** [comment]
@fduwjj yeah when I say
> But as the number of layers grow, this save should matter less and less.

I was referring to my previous sentence
> Although, combining https://github.com/pytorch/torchtitan/pull/314 with the no-folding way, we turn the beginning embedding communication from an all-reduce to a reduce-scatter.

> This folding-unfolding happens every layer

I guess you wanted to say as the number of layers grows, the save from doing collectives on dim 0 (thus avoiding `aten.cat`) happens on every layer, which I agree. (Technically speaking, the folding and unfolding only happens at the beginning and the end of an iteration, respectively.)
